lib/rules/prefer-destructuring.js

```diff
@@ -82,14 +84,39 @@ module.exports = {
         }
 
         /**
-         * Report that the given node should use destructuring
+         * Find the length to slice the nodeText and the accessor object since destructuring
+         * removes the accessor syntax from the right hand side of the AST
          *
-         * @param {ASTNode} reportNode the node to report
-         * @param {string} type the type of destructuring that should have been done
-         * @returns {void}
+         * @param {Token} lastToken the last token in the reportNode
+         * @param {SourceCode} sourceCode the source code object for simplicity
+         * @returns {Object} removalState the state from the right hand side to move/remove the text
          */
-        function report(reportNode, type) {
-            context.report({ node: reportNode, message: `Use ${type} destructuring` });
+        function findRightHandRemovalState(lastToken, sourceCode) {
```

Comment:
I think it would be better to use AST ranges to get the range that should be removed, rather than trying to parse tokens like this function does.

The right-hand side of the given node will be a `MemberExpression`:

```js
var foo = bar.baz;
```

If we can get a reference to that, we can just remove the range from `memberExpression.object.range[1]` to `memberExpression.range[1]`. That way it won't be necessary to parse the tokens inside the MemberExpression.